conda_build/jinja_context.py

```diff
@@ -203,14 +205,154 @@ def load_file_regex(config, load_file, regex_pattern, from_recipe_dir=False,
     return match if match else None
 
 
+def pin_compatible(config, package_name, upper_bound=None, pins=('p', ),
+                   permit_undefined_jinja=True):
+    """dynamically pin based on currently installed version.
+
+    only mandatory input is package_name.
+    upper_bound is the authoritative upper bound, if provided.  The lower bound is the the
+        currently installed version.
+    pin expressions can be any string separated by `.`
+        The effective pin is length of the string split on '.'.
+    """
+    compatibility = None
+
+    # this is the version split up into its component bits.
+    # There are two cases considered here (so far):
+    # 1. Good packages that follow semver style (if not philosophy).  For example, 1.2.3
+    # 2. Evil packages that cram everything alongside a single major version.  For example, 9b
+    version = get_installed_packages(config.build_prefix).get(package_name, {}).get('version')
+    if version:
+        if upper_bound:
+            compatibility = ">=" + version + ","
+            compatibility += '<{upper_bound}'.format(upper_bound=upper_bound)
+        else:
+            pins = ensure_list(pins)
+            compatibility = apply_pin_expressions(version, pins)
+
+    if not compatibility and not permit_undefined_jinja:
+        raise RuntimeError("Could not get compatibility information for {} package.  Is the "
+                            "build environment created?".format(package_name))
+    return compatibility
+
+
+# map python version to default compiler on windows, to match upstream python
+#    This mapping only sets the "native" compiler, and can be overridden by specifying a compiler
+#    in the conda-build variant configuration
+compilers = {
+    'win': {
+        'c': {
+            '2.7': 'vs2008',
+            '3.3': 'vs2010',
+            '3.4': 'vs2010',
+            '3.5': 'vs2015',
+        },
+        'cxx': {
+            '2.7': 'vs2008',
+            '3.3': 'vs2010',
+            '3.4': 'vs2010',
+            '3.5': 'vs2015',
+        },
+        'fortran': 'gfortran',
+    },
+    'linux': {
+        'c': 'gcc',
+        'cxx': 'g++',
+        'fortran': 'gfortran',
+    },
+    # TODO: Clang?  System clang, or compiled package?  Can handle either as package.
+    'osx': {
+        'c': 'gcc',
+        'cxx': 'g++',
```

Comment:
Agree completely.  I'll change it.  I need to create packages for clang and clang++ - even if they are nothing more than setting the appropriate environment variables.  Same thing for the vs20?? compilers.